File: base/sort/java/InsertionSort.java
package sort.java;

//插入排序
public class InsertionSort {

    public static void insertionSort(int[] array) {
        for (int i = 1; i < array.length; i++) {
            //前面的元素已排序，每次找后面的第一个元素在前面应该插入的位置
            int current = array[i];
            int preIndex = i - 1;
            while (preIndex >= 0 && array[preIndex] > current) {
                array[preIndex + 1] = array[preIndex];
                preIndex--;
            }
            array[preIndex + 1] = current;
        }
    }
}

File: base/sort/python/InsertionSort.py
# 时间复杂度O(n^2),最坏情况O(n^2)，最好情况O(n),空间复杂度O(1),不稳定
def insertion_sort(unsort_nums):
    for i in range(1, len(unsort_nums)):
        # 待排序的元素前一个元素的index
        pre_index = i - 1
        current = unsort_nums[i]
        while pre_index >= 0 and unsort_nums[pre_index] > current:
            # 元素向后移动一位
            unsort_nums[pre_index + 1] = unsort_nums[pre_index]
            pre_index -= 1
        # 找到插入的index后，将值插入进去
        unsort_nums[pre_index + 1] = current